common/lib/dependabot/clients/bitbucket_server.rb

```diff
@@ -0,0 +1,236 @@
+      end
+
+      def initialize(source, credential)
+        @source = source
+        @credentials = credential
+      end
+
+      def fetch_commit(_repo_name, branch)
+        path = "commits?limit=1&start=0&until=#{branch}"
+        response = get(File.join(repo_path, path))
+
+        response.body.dig("values", 0, "id")
+      end
+
+      def fetch_commit_message(branch)
+        path = "commits?limit=1&start=0&until=#{branch}"
+        response = get(File.join(repo_path, path))
+
+        response.body.dig("values", 0, "message")
+      end
+
+      def fetch_recent_commits(branch = nil)
+        path = "commits?limit=100&start=0"
+        path += "&until=#{branch}" if branch
+        response = get(File.join(repo_path, path))
+
+        response.body.dig("values")
+      end
+
+      def create_commit(branch_name, message, files)
+        files.each do |file|
+          content = Faraday::UploadIO.new(
+            StringIO.new(file.content), "application/text", file.path
+          )
+          body = {
+            content: content,
+            branch: branch_name,
+            message: message,
+            sourceCommitId: fetch_file_commit(branch_name, file.path)
+          }
+          url = File.join(repo_path, "browse", file.path)
+          put_multipart(url, body)
+        end
+      end
+
+      def fetch_file_commit(branch_name, file_path)
+        file_path = file_path.start_with?("/") ? file_path[1..-1] : file_path
+        path = "commits?limit=1&start=0&until=#{branch_name}&path=#{file_path}"
+        url = File.join(repo_path, path)
+
+        response = get(url)
+        response.body.dig("values", 0, "id")
+      end
+
+      def fetch_default_branch(_repo_name)
+        url = File.join(repo_path, "/branches/default")
+        response = get(url)
+
+        response.body.fetch("displayId")
+      end
+
+      def fetch_branch_by_name(branch)
+        url = File.join(repo_path, "/branches?filterText=#{branch}")
+        response = get(url)
+
+        response.body.dig("values")&.
+          find { |b| b["displayId"] == branch }&.
+          dig("id")
+      end
+
+      def create_branch(branch_name, base_commit)
+        url = File.join(repo_path, "/branches")
+        body = {
+          name: branch_name,
+          startPoint: base_commit
+        }
+
+        post(url, body)
+      end
+
+      # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      def fetch_repo_contents(_repo_name, commit = nil, path = nil)
+        raise "Commit is required if path provided!" if commit.nil? && path
+
+        api_path = "browse"
+        api_path += "/#{path.gsub(%r{/+$}, '')}" if path
+        api_path += "?at=#{commit}&" if commit
+        api_path += path || commit ? "limit=100" : "?limit=100"
+        response = get(File.join(repo_path, api_path))
+
+        response.body.dig("children", "values")&.map do |file|
+          file["path"] = if path.nil? || path.empty?
```

Comment:
I think there is a bug here that manifests when you specify a non-nil path.

Here's my URL:
```
[12] pry(#<Dependabot::Clients::BitbucketServer>)> File.join(repo_path, api_path)
=> "http://myserver.intranet/rest/api/1.0/projects/MYPROJECT/repos/myrepo/browse/src/webapp?at=5889d3fb62aebf564d197341b53cc3125dda505d&limit=100"
```
And here's the response (the first item in the list, anyway:

```
9] pry(#<Dependabot::Clients::BitbucketServer>)> file
=> {"path"=>{"components"=>[".gitignore"], "parent"=>"", "name"=>".gitignore", "extension"=>"gitignore", "toString"=>".gitignore"},
 "contentId"=>"fe154f85cf07bc9a2c2206344e645ceb736b9124",
 "type"=>"FILE",
 "size"=>795}
[10] pry(#<Dependabot::Clients::BitbucketServer>)> if path.nil? || path.empty?
[10] pry(#<Dependabot::Clients::BitbucketServer>)*   file.dig("path", "name")                           
[10] pry(#<Dependabot::Clients::BitbucketServer>)* else                           
[10] pry(#<Dependabot::Clients::BitbucketServer>)*   path                           
[10] pry(#<Dependabot::Clients::BitbucketServer>)* end    
```
This code says if the path is empty, join the path to the name, otherwise return the path. That doesn't sound correct - but I'm not sure what the contract for this method actually should be. @rebelagentm could you provide some more information on what this is meant to return?

Reply:
I think I've figured it out, the API returns nested objects in this form:

```
[2] pry(#<Dependabot::Clients::BitbucketServer>)> p
=> {"path"=>{"components"=>["vendor", "assets", "javascripts"], "parent"=>"vendor/assets", "name"=>"javascripts", "toString"=>"vendor/assets/javascripts"},
 "node"=>"f78cfb8ad1b8c9f9060baf6c9096062b68937f28",
 "type"=>"DIRECTORY"}
```

but the calling method only considers one level deep - if it finds a directory it makes a separate API call to find its contents. In the case of the above we just want to return "vendor", not "vendor/assets/javascripts", which was causing me a whole bunch of errors.

This all happened because I had passed the `directory` parameter to `Dependabot::Source.new`. This repo has multiple package managers so I wanted to call it a few times, once per package manager, in the specific directory.

I ended up with the following, which I think works but can probably be made more succinct:

```ruby
def fetch_repo_contents(_repo_name, commit = nil, path = nil)
  raise "Commit is required if path provided!" if commit.nil? && path

  api_path = "browse"
  api_path += "/#{path.gsub(%r{/+$}, '')}" if path
  api_path += "?at=#{commit}&" if commit
  api_path += path || commit ? "limit=100" : "?limit=100"
  response = get(File.join(repo_path, api_path))

  return response.body.dig("children", "values").map do |p|
    f = {
      "type" => p["type"]
    }
    if p["path"]["components"].nil?
      f["path"] = p["path"]["toString"]
    else
      f["path"] = p["path"]["components"][0]
    end
    f
  end
end
```